pkg/handlers/handlers.go

```diff
@@ -1,44 +1,99 @@
 package handlers
 
 import (
+	"errors"
 	"fmt"
 	"io"
 
 	"github.com/gabriel-vasile/mimetype"
+	"github.com/mholt/archiver/v4"
 
 	logContext "github.com/trufflesecurity/trufflehog/v3/pkg/context"
 	"github.com/trufflesecurity/trufflehog/v3/pkg/readers"
 	"github.com/trufflesecurity/trufflehog/v3/pkg/sources"
 )
 
-// readSeekCloser is an interface that combines the functionality of io.ReadSeekCloser and io.ReaderAt.
-// It supports reading data, seeking within an open resource, and closing the resource once operations are complete.
-// Additionally, it allows reading from a specific offset within the resource without altering its current position,
-// enabling efficient and flexible data access patterns. This interface is particularly useful for handling files
-// or other data streams where random access and sequential processing are required.
-type readSeekCloser interface {
-	io.ReadSeekCloser
-	io.ReaderAt
+// fileReader is a custom reader that wraps an io.Reader and provides additional functionality for identifying
+// and handling different file types. It abstracts away the complexity of detecting file formats, MIME types,
+// and archive types, allowing for a more modular and extensible file handling process.
+//
+// fileReader leverages the archiver and mimetype packages for file type identification and provides information
+// about the detected file format, MIME type, and whether the file is an archive. This information can be
+// used by FileHandler implementations to make decisions on how to process the file.
+//
+// By encapsulating the file type detection logic, fileReader simplifies the implementation of FileHandler and
+// promotes a more cohesive and maintainable codebase. It also embeds a BufferedFileReader to provide efficient
+// random access to the file content.
+type fileReader struct {
+	format   archiver.Format
+	mimeType mimeType
+	*readers.BufferedFileReader
+	isArchive bool
+}
+
+func newFileReader(r io.ReadCloser) (fileReader, error) {
+	defer r.Close()
+
+	var (
+		reader fileReader
+		rdr    *readers.BufferedFileReader
+		err    error
+	)
+	rdr, err = readers.NewBufferedFileReader(r)
+	if err != nil {
+		return reader, fmt.Errorf("error creating random access reader: %w", err)
+	}
+	reader.BufferedFileReader = rdr
+
+	// Ensure the reader is closed if an error occurs after the reader is created.
+	// During non-error conditions, the caller is responsible for closing the reader.
+	defer func() {
+		if err != nil && rdr != nil {
+			_ = rdr.Close()
+		}
+	}()
+
+	format, arReader, err := archiver.Identify("", rdr)
+	switch {
+	case err == nil: // Archive detected
+		reader.isArchive = true
```

Comment:
I find this field name confusing because it's only set to true for certain types of archives. I'm not sure what a better one would be, though - maybe `isGenericArchive` or `isNonSpecializedArchive` or something like that?

Reply:
That's a valid point. I will rename it to `isGenericArchive`.